v3_async_wip/v3_async_wip/sastoken.py

```diff
@@ -5,173 +5,222 @@
             # Because of variant signing mechanisms, we don't know what error might be raised.
             # So we catch all of them.
-            raise SasTokenError("Unable to build SasToken from given values") from e
+            raise SasTokenError("Unable to generate SasToken") from e
         url_encoded_signature = urllib.parse.quote(signature, safe="")
-        if self._key_name:
-            token = self._auth_rule_token_format.format(
-                resource=url_encoded_uri,
-                signature=url_encoded_signature,
-                expiry=str(self.expiry_time),
-                keyname=self._key_name,
-            )
-        else:
-            token = self._simple_token_format.format(
-                resource=url_encoded_uri,
-                signature=url_encoded_signature,
-                expiry=str(self.expiry_time),
-            )
-        return token
+        token_str = TOKEN_FORMAT.format(
+            resource=url_encoded_uri,
+            signature=url_encoded_signature,
+            expiry=str(expiry_time),
+        )
+        return SasToken(token_str)
 
-    @property
-    def expiry_time(self) -> int:
-        """Expiry Time is READ ONLY"""
-        return self._expiry_time
 
+class ExternalSasTokenGenerator(SasTokenGenerator):
+    def __init__(self, generator_fn: Union[Callable[[], str], Callable[[], Awaitable[str]]]):
+        """An object that can generate SasTokens by invoking a provided callable.
+        This callable can be a function or a coroutine function.
 
-class NonRenewableSasToken(SasToken):
-    """NonRenewable Shared Access Signature Token used to authenticate a request.
+        :param generator_fn: A callable that takes no arguments and returns a SAS Token string
+        :type generator_fn: Function or Coroutine Function which returns a string
+        """
+        self.generator_fn = generator_fn
 
-    This token is 'non-renewable', which means that it is invalid once it expires, and there
-    is no way to keep it alive. Instead, a new token must be created.
+    async def generate_sastoken(self) -> SasToken:
+        try:
+            # NOTE: the typechecker has some problems here, so we help it with a cast.
+            if asyncio.iscoroutinefunction(self.generator_fn):
+                generator_fn = cast(Callable[[], Awaitable[str]], self.generator_fn)
+                token_str = await generator_fn()
+            else:
+                generator_coro_fn = cast(Callable[[], str], self.generator_fn)
+                token_str = generator_coro_fn()
+            return SasToken(token_str)
+        except Exception as e:
+            raise SasTokenError("Unable to generate SasToken") from e
 
-    Data Attributes:
-    expiry_time (int): Time that token will expire (in UTC, since epoch)
-    resource_uri (str): URI for the resource the Token provides authentication to access
-    """
 
-    def __init__(self, sastoken_string) -> None:
+class SasTokenProvider:
```

Comment:
I had talked about calling this a SasTokenWarden, but ultimately, I stuck with SasTokenProvider due to semantics - while I think Warden is a better name for what it does, from the perspective of something using this class, Warden is meaningless, and reduces clarity. A user of this class uses it to provide the token - whatever the wholistic picture of what is happening inside is irrelevant.